File: test/ShoppingApp.Tests/InventoryTests.cs
using Orleans.TestingHost;
using ShoppingApp.Abstractions;
using ShoppingApp.Tests.Initialization;
using Xunit;

namespace ShoppingApp.Tests;

[Collection(ClusterCollection.ClusterFixtureName)]
public class InventoryTests(ClusterFixture clusterFixture)
{
	private TestCluster Cluster { get; } = clusterFixture.Cluster;

	[Fact]
    public async Task VerifyingSuccessfulAdditionAndRetrievalOfProduct()
	{
		// Arrange
		var grain = Cluster.GrainFactory.GetGrain<IInventoryGrain>(string.Empty);

		// Act
		await grain.AddOrUpdateProductAsync(new ProductDetails { Id = "1", Name = "Test" });
		var products = await grain.GetAllProductsAsync();

		// Assert
		Assert.True(products.Count == 1);
    }
}

File: test/ShoppingApp.Tests.FSharp/InventoryTests.fs
module Tests

open Initialization
open ShoppingApp.Abstractions
open System
open Xunit
open Xunit.Abstractions

[<Collection(CurrentAssembly.ClusterFixture)>]
type InventoryTests(clusterFixture: ClusterFixture, output: ITestOutputHelper) =
    let cluster = clusterFixture.Cluster
    
    [<Fact>]
    member _.``Verifying Successful Addition and Retrieval of a Product``() =
        async {
            // Arrange
            let grain = cluster.GrainFactory.GetGrain<IInventoryGrain>(nameof(IInventoryGrain))
            
            // Act
            do! grain.AddOrUpdateProductAsync(new ProductDetails(Id = "1", Name = "Test")) |> Async.AwaitTask
            let! products = grain.GetAllProductsAsync() |> Async.AwaitTask
            output.WriteLine(products.Count.ToString())
            
            // Assert
            Assert.True(products.Count = 1)
    }